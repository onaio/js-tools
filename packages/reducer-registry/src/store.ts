/** Store module */
import { configureStore } from '@reduxjs/toolkit';
import { combineReducers, compose, createStore, Reducer } from 'redux';
import reducerRegistry, { Registry } from './registry';

/** Declare type for initial state */
interface State {
  [key: string]: any;
}

/** Add redux dev tools to Window */
declare global {
  interface Window {
    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__?: typeof compose;
  }
}

/** Combine all reducers, but preserve any initial state
 */
export function combine(reducers: Registry, initialState: State = {}) {
  Object.keys(initialState).forEach(item => {
    reducers[item] = (state = initialState[item]): Reducer => state;
  });
  return combineReducers(reducers);
}

/** Function that returns a Redux store given a list of Reducers and initial
 * state
 */
// export function getStore(reducers: Registry, initialState: State = {}) {
//   if (Object.keys(reducers).length > 0) {
//     return configureStore({
//       reducer: combine(reducers, initialState)
//     });
//   }
//   return configureStore({
//     reducer: (() => initialState)
//   });
// }

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export function getStore(reducers: Registry, initialState: State = {}) {
  if (Object.keys(reducers).length > 0) {
    return configureStore({
      reducer: combine(reducers, initialState)
    });
  }
  return configureStore({
    reducer: () => initialState
  });
}

// export function getStore(reducers: Registry, initialState: State = {}) {

//   return configureStore({
//     reducer: combine(reducers, initialState),
//     preloadedState: initialState,
//     devTools: process.env.NODE_ENV !== 'production',
//     enhancers: [composeEnhancers()],
//   });
// }

/** Ready-to-use default store made from an empty Reducer registry */
const store = getStore(reducerRegistry.getReducers());

/** Add reducer to store when it is registered */
reducerRegistry.setChangeListener(reducers => {
  store.replaceReducer(combine(reducers));
});

export default store;
