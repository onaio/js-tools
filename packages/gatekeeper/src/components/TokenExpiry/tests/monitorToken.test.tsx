import store from '@onaio/connected-reducer-registry';
import reducerRegistry from '@onaio/redux-reducer-registry';
import session, { reducerName as sessionReducer, updateTokenExpired } from '@onaio/session-reducer';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { createBrowserHistory } from 'history';
import React from 'react';
import { Provider } from 'react-redux';
import { Router } from 'react-router';
import ConnectedMonitorTokenExpiry, { MonitorTokenExpiry } from '../monitorToken';

reducerRegistry.register(sessionReducer, session);
const history = createBrowserHistory();
const redirectUrl = '/session/expired';

describe('/components/errors/Fallback', () => {
  it('renders without crasshing', () => {
    shallow(<MonitorTokenExpiry redirectUrl={redirectUrl} />);
  });

  it('render correctly', () => {
    const wrapper = mount(
      <Router history={history}>
        <MonitorTokenExpiry redirectUrl={redirectUrl} />
      </Router>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('render correctly when connected to store', () => {
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <ConnectedMonitorTokenExpiry redirectUrl={redirectUrl} />
        </Router>
      </Provider>
    );
    expect(window.location.pathname).toEqual('/');
    // update token expiry status
    store.dispatch(updateTokenExpired(true));
    wrapper.update();

    expect(window.location.pathname).toEqual(redirectUrl);
  });
});
