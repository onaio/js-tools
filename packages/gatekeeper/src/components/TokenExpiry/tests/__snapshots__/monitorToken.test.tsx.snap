// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/components/errors/Fallback render correctly 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <MonitorTokenExpiry
    redirectUrl="/session/expired"
    tokenExpired={false}
  />
</Router>
`;
